### PR TITLE
Pubsub

### PR DESCRIPTION
**Pull request by JamesTFarrington**
Pubsub tests now pass consistently.

### CODE REVIEW

**Comment by JamesTFarrington** [comment]
tox result:
===============================================================================
[SKIPPED]
Pending new logic.

tiempo.tests.test_scheduling.TaskScheduleTests.test_minutely_tasks_expire_in_60_seconds
===============================================================================
[FAIL]
Traceback (most recent call last):
  File "/home/james/reelio/tiempo/tiempo/tests/test_job_reporting.py", line 53, in check
    self.assertTrue(cj)
  File "/home/james/reelio/tiempo/.tox/py27/local/lib/python2.7/site-packages/twisted/trial/_synctest.py", line 400, in assertTrue
    super(_Assertions, self).assertTrue(condition, msg)
  File "/usr/lib/python2.7/unittest/case.py", line 422, in assertTrue
    raise self.failureException(msg)
twisted.trial.unittest.FailTest: OrderedDict() is not true

tiempo.tests.test_job_reporting.JobReportingTests.test_show_some_completed_jobs
-------------------------------------------------------------------------------
Ran 45 tests in 7.390s

FAILED (skips=1, failures=1, successes=43)

File: tiempo/conn.py [hunk 1]
```diff
@@ -19,11 +20,17 @@ def subscribe_to_backend_notifications(db=REDIS_QUEUE_DB):
...
 
 
-def hear_from_backend():
-    events = []
-    def parse_backend():
+def create_event_queue():
+    """
+    Returns a closure that updates and returns the events queue
+    """
+    events = deque([])
```
Comment (JamesTFarrington) [inline tiempo/conn.py]: I feel that queues are better for this use case, I also change the name of the function, and changed it into a closure.

File: tiempo/conn.py [hunk 2]
```diff
@@ -32,4 +39,18 @@ def parse_backend():
...
         return events
-    return parse_backend
+    return update_event_queue
+
+def check_backend():
+    """
+    A syncronous alternative for update_event_queue.
+    Parses all events and returns them as a list.
+    """
+    events = []
```
Comment (JamesTFarrington) [inline tiempo/conn.py]: Added hear_from_backend back with a new name for developer convenience.

File: tiempo/runner.py [hunk 3]
```diff
@@ -216,3 +200,15 @@ def shut_down(self):
...
+    """
+    A callback for runner management.
+
+    Creates a new runner and shutsdown the old one.
+    """
+    number = runner.number
+    task_groups = runner.task_groups
+    new_runner = Runner(number, task_groups)
+    runner.shut_down
+    return new_runner
```
Comment (JamesTFarrington) [inline tiempo/runner.py]: rewrote cleanup as a function.

File: tiempo/tests/test_loop_cycle.py [hunk 4]
```diff
@@ -21,11 +26,31 @@ def unblocker():
...
+    def teardown(self):
+        self.looper.stop()
+        REDIS.flushall()
+
+    def test_glean_returns_event_queue(self):
+        events = glean_events_from_backend()
+        self.assertIsInstance(events, deque)
+
+    def test_whole_cycle(self):
+        self.fail()
```
Comment (JamesTFarrington) [inline tiempo/tests/test_loop_cycle.py]: Starting to implement tests for cycle, which we will need for tiempo to be finished.

File: tiempo/tests/test_pubsub.py [hunk 5]
```diff
@@ -33,12 +36,38 @@ def test_result_is_properly_reported(self):
...
+        REDIS.flushall()
+        subscribe_to_backend_notifications()
+        REDIS.config_get('notify-keyspace-events')
+        time.sleep(.1)
+
+        subscribe_event = check_backend()[0]
+        self.assertEqual(subscribe_event['type'], 'psubscribe')
+
+        REDIS.set('results:whatever', 'tapestry needle')
+
```
Comment (JamesTFarrington) [inline tiempo/tests/test_pubsub.py]: I rewrote the test for update_queue, then added the old test back in when I added check_backend.

File: tiempo/work.py [hunk 6]
```diff
@@ -348,6 +348,49 @@ def __setup(self, args):
...
+        runs the callable.
+
+        All paths of execution should eventually lead here.
+        """
+
+        func = self.get_function()
+        kwargs = getattr(self, 'kwargs_to_function', {})
+
+        if self.announcer_name:
+            kwargs[self.announcer_name] = runner.announcer
```
Comment (JamesTFarrington) [inline tiempo/work.py]: Moved run and get_function closer to the top and added docstrings; to emphasis their importance and make them easier to understand.

File: tiempo/work.py [hunk 7]
```diff
@@ -679,7 +688,7 @@ def spawn_job_and_run_now(self, *args, **kwargs):
     def soon(self, tiempo_wait_for=None,
              *args, **kwargs):
         '''
-        Just like spawn_job_and_run_roon(), but returns self instead of the job.
+        Just like spawn_job_and_run_soon(), but returns self instead of the job.
```
Comment (JamesTFarrington) [inline tiempo/work.py]: fixed a typo

**Comment by jMyles** [comment]
Hey man.  As you can see, 2.0-thunderdome is a rebase.  Can you please rebase your work here against c3211183d1c5bbb05399ec06bbba2b619c75e1c9 (your 65fd35b4fcba54df088828b7425aea389b5416b4)?  I want to only see the commits *after* ```Returns handle_message changes error is handled in parse_backend```

**Pull request closed without merging**